### ISSUE TITLE
Routes with one required and one optional parameter (id + slug)

### ISSUE THREAD

**Issue by tillkolter** [nuxt/nuxt.js#646]
In my currenct project I was using a route with a dynamic pattern which constituted of two params (`releases/:id-:slug`) where the slug part was generated only for SEO purposes.

I am now trying to migrate to Nuxt (since any SEO optimization is useless, if the page is not rendered server-side), but I have no idea how to create a folder structure which generates that router pattern without duplicating the source code (one `_id/index.vue` file for the required :id param and a `_id/_slug.vue` file for the optional slug parameter which both would have the same ). I tried to programmatically extend the routes which works for nuxt-links but since the standard route is generated anyway, when I refresh the page, the dynamic part is assumed to be the `id` part 

Is it possible to migrate my pattern or should I extend my database to retrieve data directly through the slug?

**Comment by alexchopin** [comment]
This folder structure should works:

```bash
pages/
- releases/
-- index.vue
-- _id/
--- _slug.vue
```

`pages/releases/index.vue` will be transformed as this route: `/releases` (should be the view for all the releases)
`pages/releases/_id/_slug.vue` will be transformed as this route: `/releases/:id/:slug?` (should be the view for a release where ID is required and SLUG is optional)

**Comment by tillkolter** [comment]
@alexchopin I am afraid that did not answer my question. Let me rephrase it.

My slug parameter is not just optional. It should not to even change anything concerning the content of the page hence the source code of both files would be the same. The semantic url is just meant to improve SEO ranking, but the retrieval of the database entry depends strictly on the id part of the url. 

I was able to realize this by constructing a regular vue router programmatically but extending the router in the nuxt.config.js doesnt fully solve the case since the automatic generated 'releases-id' (releases/:id) route conflicts then with the programmatic 'releases/:id-:slug' version as 'releases/123-dj-awesome-mixtape-2' would be recognised with :id='123-dj-awesome-mixtape-2'.

From what I could tell after experimenting with Nuxt for 3 days, I think this is not possible, but I was hoping to get a hint how this could be solved.

For certain reasons I could not change the schema of the original releases table and I wanted to avoid creating another table only to store the related slug field which is why I wanted to rely on the id only while still generating semantic urls. Besides that, the slug would not even be reliable to be unique. But I will now go down that road if the wanted url pattern can not be realized.

**Comment by alexchopin** [comment]
Can you post your `extendRoutes` function to help you to improve it?

**Comment by tillkolter** [comment]
@alexchopin 
Here is what I wanted to use:

```
router: {
    extendRoutes (routes, resolve) {
      routes.push(
        {
          name: 'genres-id-slug',
          path: '/genres/:id-:slug,
          component: resolve(__dirname, 'pages/genres/_id/index.vue')
        }
      )
    }
}
```

But in this case the default route (`genres/_id/index.vue --> genres-id`) is standing in the way of parsing the pattern `genres/:id-:slug` in favour of `genres/:id`. I guess there is no clean work around, so I accepted that fact already and decided to touch my database and let the slugs be the key for retrieval.

In any case, I had to extend my routes, because I have a genre <-> subgenre relation which both types rendering to the same page structure (with just the tiny difference, that the super genre page renders a pulldown menu to select a subgenre). I didnt want to duplicate my pages, because the rendering difference is just a `v-if` check, if the genre has any subgenres. So I came up with the solution to create another route which points to the same page. I dont know if this was the best solution, but it worked.

```
routes.push({
          name: 'genres-slug-subslug',
          path: '/genres/:slug/:subslug',
          component: resolve(__dirname, 'pages/genres/_slug/index.vue')
})
```

I don't know if it makes sense to keep such cases in mind for future versions of the project, it's just meant to be food for thought.

Btw: I am so happy with migrating my project to Nuxt and finally have the rendered pages to bring an appropriate Facebook sharing to life. Good work!
